components/candle/masteryService.ts

```diff
@@ -75,52 +75,78 @@ export class MasteryService {
         }
     }
 
-    getCompletionData(
-        locationParentId: string,
-        subLocationId: string,
-        gameVersion: GameVersion,
+    /**
+     * Get generic completion data stored in a user's profile. Called by both `getLocationCompletion` and `getFirearmCompletion`.
+     * @param userId The id of the user.
+     * @param gameVersion The game version.
+     * @param completionId An Id used to look up completion data in the user's profile. Can be `parentLocationId` or `progressionKey`.
+     * @param maxLevel The max level for this progression.
+     */
+    private getCompletionData(
         userId: string,
-    ): CompletionData {
-        if (!this.masteryData.has(locationParentId)) {
-            return undefined
-        }
-
-        //Get the mastery data
-        const masteryData: MasteryPackage =
-            this.masteryData.get(locationParentId)
-
+        gameVersion: GameVersion,
+        completionId: string,
+        maxLevel: number,
+    ) {
         //Get the user profile
         const userProfile = getUserData(userId, gameVersion)
 
-        //Gather all required data
-        const lowerCaseLocationParentId = locationParentId.toLowerCase()
-
-        userProfile.Extensions.progression.Locations[
-            lowerCaseLocationParentId
-        ] ??= {
+        // Generate default completion before trying to acquire it
+        userProfile.Extensions.progression.Locations[completionId] ??= {
```

Comment:
This is what I mean with the remark about the completedId being sketchy. A sniper level might use a progression key like `FIREARMS_SC_HERO_SNIPER_HM`, so won't this add non-existing locations to the player profile, for each sniper in the game? Also, what if two locations have the same sniper for its progressionKey, you will get progress of the wrong location, no?

Reply:
> won't this add non-existing locations to the player profile, for each sniper in the game?
Exactly. For sniper missions, mastery is tied to the sniper rifles, and I just re-used the data structure that stores normal mastery progression. All nine snipers have unique `progressionKey`s and I think that was intended by IOI by calling it a "key".